cpp/src/arrow/type.h

```diff
@@ -136,7 +136,10 @@ struct Type {
     MAP,
 
     /// Custom data type, implemented by user
-    EXTENSION
+    EXTENSION,
+
+    /// Incomplete dictionary type
+    INCOMPLETE_DICTIONARY
```

Comment:
See https://issues.apache.org/jira/browse/ARROW-3144.

Before doing more detailed review, having a new type enum here is one option (which introduces some development complexity). Another option is to have `IncompleteDictionary` be a subclass of `DictionaryType` but otherwise identify itself as DICTIONARY.

I had originally thought to have a _mutable_ dictionary type with atomic mutations to permit delta dictionaries to update "connected" arrays. What do you think about that?

Reply:
So `DictionaryType` would gain a virtual method to indicate whether or not the dictionary is known yet, but also allowing for the dictionary to grow through deltas

Reply:
Hmm... I'm uncomfortable with the idea of a mutable data type, especially if it can change in another thread... The benefit of a dedicated enum value is that it forces us to think about both situations (a dictionary with known values, a dictionary with unknown values).

As for delta dictionaries, I don't know. Is the "delta" something that needs to be represented at the data type level, or is it a mere array? If the latter, we simply need to expose an operation to build a new `DictionaryType` with concatenated values, AFAICT.

(right now there's support for delta dictionaries in `DictionaryBuilder`, though it's a bit ad hoc)

@xhochy @emkornfield Any opinions?